module/x/peggy/abci.go

```diff
@@ -20,50 +20,49 @@ func slashing(ctx sdk.Context, k keeper.Keeper) {
 	params := k.GetParams(ctx)
 	currentBondedSet := k.StakingKeeper.GetBondedValidatorsByPower(ctx)
 
-	// valsets are sorted so the most recent one is first
-	valsets := k.GetValsets(ctx)
-	if len(valsets) == 0 {
-		k.SetValsetRequest(ctx)
-	}
+	// valsets are sorted by nonce in ASC order
+	unslashedValsets := k.GetUnSlashedValsets(ctx)
 
-	for i, vs := range valsets {
+	for _, vs := range unslashedValsets {
 		signedWithinWindow := uint64(ctx.BlockHeight()) > params.SignedValsetsWindow && uint64(ctx.BlockHeight())-params.SignedValsetsWindow > vs.Height
-		switch {
+		if !signedWithinWindow {
```

Comment:
we can get rid of this once we make the change from the comment below. We'll still need to separately get the current valset and check it

Reply:
we should ensure we prevent overflow however. Something like 
```
    slashingBlockHeightThreshold := uint64(ctx.BlockHeight()) - params.SignedValsetsWindow
    // don't slash in the beginning before there aren't even SignedValsetsWindow blocks yet
    if uint64(ctx.BlockHeight()) > params.SignedValsetsWindow {
        slashingBlockHeightThreshold = uint64(0)
    }
```